Source/VisualStudioBlueprintDebuggerHelper/Private/VisualStudioBlueprintDebuggerHelperModule.cpp

```diff
@@ -0,0 +1,245 @@
+
+#include "VisualStudioBlueprintDebuggerHelperModule.h"
+#include <Modules/ModuleManager.h>
+#include <UObject/Script.h>
+#include <UObject/Stack.h>
+#include <UObject/Object.h>
+#include <Templates/Casts.h>
+#include <Kismet2/KismetDebugUtilities.h>
+#include <Containers/Array.h>
+#include <UObject/Class.h>
+#include <Engine/BlueprintGeneratedClass.h>
+#include <Engine/Blueprint.h>
+#if ENGINE_MAJOR_VERSION >= 5 && ENGINE_MINOR_VERSION >= 4
+#include <Blueprint/BlueprintExceptionInfo.h>
+#endif
+#include <Internationalization/Text.h>
+#include <HAL/Platform.h>
+#include <EdGraph/EdGraphNode.h>
+#include <EdGraph/EdGraphPin.h>
+#include <Templates/SharedPointer.h>
+#include <Templates/Tuple.h>
+#include <CoreGlobals.h>
+#include <map>
+
+IMPLEMENT_MODULE(FVisualStudioBlueprintDebuggerHelper, VisualStudioBlueprintDebuggerHelper);
+
+DEFINE_LOG_CATEGORY(LogVisualStudioBlueprintDebuggerHelper);
+
+#if ENGINE_MAJOR_VERSION >= 5
+#define FCustomBlueprintPropertyInfo TSharedPtr<FPropertyInstanceInfo>
+#else
+#define FCustomBlueprintPropertyInfo FDebugInfo
+#endif
+
+struct FVSNodePinRuntimeInformation
+{
+	UEdGraphPin* Pin;
+	FCustomBlueprintPropertyInfo Property;
+
+	FVSNodePinRuntimeInformation(UEdGraphPin* InPin, FCustomBlueprintPropertyInfo InProperty)
+		: Pin(InPin)
+		, Property(InProperty)
+	{
+	}
+};
+
+struct FVSNodeData
+{
+	FText NodeName;
+	TArray<TSharedPtr<FVSNodePinRuntimeInformation>> Properties;
+	int32 ScriptEntryTag;
+	const UEdGraphNode* Node;
+};
+
+struct FVSNodesRuntimeInformation
+{
+	TArray<TSharedPtr<FVSNodeData>> Nodes;
+};
+
+struct FVSBlueprintRuntimeInformation
+{
+	TArray<TTuple<UBlueprint*, TSharedPtr<FVSNodesRuntimeInformation>>> RunningBlueprints;
+};
+
+struct StackTraceHelper
+{
+	int32 ScriptEntryTag;
+	FString NodeName;
+};
+
+// Keep exported so we can read it.
+VISUALSTUDIOBLUEPRINTDEBUGGERHELPER_API FVSBlueprintRuntimeInformation BlueprintsRuntimeInformation;
+
+VISUALSTUDIOBLUEPRINTDEBUGGERHELPER_API std::map<void*, StackTraceHelper> StackFrameInformation;
+
+VISUALSTUDIOBLUEPRINTDEBUGGERHELPER_API const char* DebuggerHelperVersion = "1.0.0";
+
+void FVisualStudioBlueprintDebuggerHelper::StartupModule()
+{
+	CurrentScriptEntryTag = 0;
+
+	FBlueprintContextTracker::OnEnterScriptContext.AddRaw(
+		this,
+		&FVisualStudioBlueprintDebuggerHelper::OnEnterScriptContext);
+
+	FBlueprintContextTracker::OnExitScriptContext.AddRaw(
+		this,
+		&FVisualStudioBlueprintDebuggerHelper::OnExitScriptContext);
+
+	FBlueprintCoreDelegates::OnScriptException.AddRaw(
+		this,
+		&FVisualStudioBlueprintDebuggerHelper::OnScriptException);
+}
+
+void FVisualStudioBlueprintDebuggerHelper::ShutdownModule()
+{
+	FBlueprintCoreDelegates::OnScriptException.RemoveAll(this);
+	FBlueprintContextTracker::OnExitScriptContext.RemoveAll(this);
+	FBlueprintContextTracker::OnEnterScriptContext.RemoveAll(this);
+}
+
+void FVisualStudioBlueprintDebuggerHelper::OnEnterScriptContext(
+	const struct FBlueprintContextTracker& Context,
```

Comment:
just curious, is it some requirement of UE coding standard?
Haven't seen explicit `struct ...` in arguments type names in C++ code for a long time, I bet it's a C legacy where it was required :)

Reply:
I could not find any guidelines about it. However, in this specific case it's just a object to hold data and no logic is provide. So I guess they just wanted to make it "all public". But this is just a guess...